app/controllers/application_controller.rb

```diff
@@ -3,8 +3,148 @@
+  end
     set :public_folder, 'public'
     set :views, 'app/views'
+
+    get '/' do
+      erb :index
+    end
+
+    def logged_in?
+      !!session[:user_id]
+    end
+
+    get '/signup' do
+      if self.logged_in?
+          redirect "/tweets"
+      else
+        erb :'/users/create_user'
+      end
+    end
+
+    post '/signup' do
+      if params[:username]=="" || params[:email]==""|| params[:password]==""
+        redirect "/signup"
+      else
+        @user = User.create(params)
+        session[:user_id] = @user.id
+        redirect "/tweets"
+      end
+    end
+
+    get '/login' do
+        if self.logged_in?
+          redirect "/tweets"
+        else
+          erb :'/users/login'
+        end
+    end
+
+    post '/login' do
+      @user = User.find_by(username: params[:username])
+      if @user && @user.authenticate(params[:password])
+        session[:user_id] = @user.id
+        redirect "/tweets"
+      else
+        redirect "/login"
+      end
+
+    end
+
+    get '/logout' do
+      if self.logged_in?
+        session.clear
+        redirect "/login"
+      else
+        redirect "/"
+      end
+    end
+
+    get '/tweets' do
+
+    if logged_in?
+      @tweets = Tweet.all
+      @user = User.find(session[:user_id])
+      erb :'/tweets/tweets'
+    else
+      redirect "/login"
+    end
   end
 
-end
\ No newline at end of file
+    get '/users/:slug' do
+      @user = User.find_by_slug(params[:slug])
+      erb :'/users/show'
+    end
+
+    get '/tweets/new' do
+      if self.logged_in?
+        erb :'/tweets/create_tweet'
+      else
+        redirect "/login"
+      end
+      end
+
+    post '/tweets/new' do
+      @user = User.find(session[:user_id])
+      if self.logged_in? && params[:content] != ""
+        @user.tweets << Tweet.new(content: params[:content])
+        redirect "/users/#{@user.slug}"
+      else
+        redirect "/tweets/new"
+      end
+    end
+
+    get '/tweets/:id/edit' do
+      if logged_in?
+        @tweet = Tweet.all.find(params[:id])
+        if @tweet && session[:user_id] == @tweet.user_id
+        erb :'tweets/edit_tweet'
+        else
+          redirect "/tweets"
+        end
+      else
+        redirect "/login"
+      end
+    end
+
+    patch '/tweets/:id' do
+      if logged_in?
+        @tweet = Tweet.all.find(params[:id])
+        if session[:user_id] == @tweet.user_id && params[:content] != ""
+          @tweet.update(content: params[:content])
+          redirect "/tweets/#{@tweet.id}"
+        else
+          redirect "/tweets/#{@tweet.id}/edit"
+        end
+      else
+        redirect "/login"
+      end
+    end
+
+    get '/tweets/:id' do
+      @tweet = Tweet.all.find(params[:id])
+      if logged_in?
+        erb :'tweets/show_tweet'
+      else
+        redirect "/login"
+      end
+    end
+
+
+  post '/tweets/:id/delete' do
+    @tweet = Tweet.all.find(params[:id])
+    if logged_in? && session[:user_id] == @tweet.user_id
```

Comment:
Prefer to find tweets that are already only associated to the `current_user` than to manually check the `user_id` foreign key of the tweet against the `current_user.id`

```ruby
tweet = current_user.tweets.find_by(id: params[:id])
if tweet && tweet.destroy
  redirect "/tweets"
else
  redirect "/tweets/#{tweet.id}"
end
```

We're using the association collection method provided by [`has_many`](http://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html#method-i-has_many) to query the user's tweets. 

Note the usage of [`find_by`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find-by) vs [`find`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find). Had we used `find` here, ActiveRecord would raise an error in the event that we could not find a tweet associated with this user. `find_by` will allow this action to fail gracefully, simply returning nil if the user is trying to delete a tweet that they do not have access to.